- + 연산자
    
    문자열에는 + 연산자로 문자열 연결 연산을 적용할 수 있다.
    
    ex) “가나다” + “라마바” → 가나다라마바
    
    But 문자열은 무조건 문자열끼리 + 연산자를 사용해서 연결해야 하므로, 문자인 숫자? 를 문자열과 더하고 싶다면 큰따옴표를 붙여 문자열로 인식시킨 후에만 원하는 결과를 얻을 수 있다.
    
    “안녕하세요” + 1 → ERROR
    
    “안녕하세요” + “1” → 안녕하세요1

- * 연산자
    
    문자열을 숫자와 * 연산자로 연결하면 문자열을 반복할 수 있다.
    
    ex) print(”안녕하세요” * 3) → 안녕하세요안녕하세요안녕하세요

문자열의 특정 범위를 선택할 때 사용하는 연산자.

ex) print(”안녕하세요”[1:4]) → 녕하세 

→ [a:b] 를 사용하게 된다면, a 부터 b-1까지 임을 잊지 않아야 한다.

헷갈리지 말 것

배열의 시작이 0인 것처럼, [1] 은 [0] 다음이므로, “**안”**이 아니라 “**녕”**인 것이다.

---

ex) print(”안녕하세요”[:4]) → 안녕하세 

→ [0:4] 에서 0이 생략되었음을 알 수 있다.

---

ex) print(”안녕하세요”[2:]) → 하세요

→ [2:5]에서 5가 생략되었음을 알 수 있다.

IndexError 예외는 리스트/문자열의 수를 넘는 요소/글자를 선택할 때 발생한다. 

쉽게 예시를 들자.

print(”안녕하세요”[6]) 이라는 코드를 작성하게 된다면, 문자열의 길이를 넘는 부분을 선택했기 때문에 IndexError가 발생하게 된다.

/과 다르게 // 연산자를 사용한 수식의 결과는 소수점 아래를 떼어 버린 값이 출력되는 것을 확인할 수 있다. 

서로 다른 자료를 연산하면 TypeError 라는 예외가 발생한다.

In [1]:
string1 = input()
print(string1)

gdgd


문자열을 숫자로 변환하는 과정을 CAST 라고 한다. int 를 제외하고 float를 사용하면 실수형으로 변경할 수 있다.

In [8]:
string_a = input("입력하세요")
int_a = int(string_a)

print("입력받은 string_a = " + string_a)
print("int_a의 타입은 : " + str(type(int_a)))
print("string_a의 타입은 : " + str(type(string_a)))

입력받은 string_a = 12345
int_a의 타입은 : <class 'int'>
string_a의 타입은 : <class 'str'>


자료형을 변환할 때 **변환할 수 없는 것을 변환하려고 하면** ValueError 예외가 발생한다. 

예를 들어, **숫자가 아닌 것을 숫자로 변환하려고 할 때**와 같은 상황이 존재한다. 

다른 예시로는 소수점이 있는 숫자 형식의 문자열을 int() 함수로 변환하려고 할 때이다.

In [10]:
string_a = input("소수점이 있는 숫자를 입력하시오")
float_a = float(string_a)

print("입력받은 string_a = " + string_a)
print("float_a의 타입은 : " + str(type(float_a)))
print("string_a의 타입은 : " + str(type(string_a)))

print("소수점이 들어간 숫자를 int 형식으로 변환하면 ValueError가 날 것이다.")
print(int(string_a))
      


입력받은 string_a = 123.12
float_a의 타입은 : <class 'float'>
string_a의 타입은 : <class 'str'>
소수점이 들어간 숫자를 int 형식으로 변환하면 ValueError가 날 것이다.


ValueError: invalid literal for int() with base 10: '123.12'

format() 함수는 문자열이 가지고 있는 함수로, 중괄호를 포함한 문자열 뒤에 마침표를 찌ㄱ고 format() 함수를 사용하는데, 중괄호의 개수와 .format 함수 괄호 안 매개변수의 개수는 반드시 같아야 한다.

In [11]:
string_a = "{}".format(10)

print(string_a)
print(type(string_a))

10
<class 'str'>


In [21]:
format_c = "{} {} {} {} {}".format("나는", "사과", 10, "개를 먹고싶어", True)
print(format_c)

나는 사과 10 개를 먹고싶어 True


1. format 함수에서 d 앞에 + 기호를 붙이면 양수와 음수 기호를 표현할 수 있다. 
2. format 함수에서 d 앞에 숫자를 붙이면 그 만큼의 공백을 만든 후 그 뒤에 숫자를 채워준다.
3. 혹은 기호와 공백을 조합할 수 있는데, 이 때는 = 기호를 앞에 붙여서 사용할 수 있다.
4. 또는 +0을 넣어줌으로써 공백이 생긴 부분을 0으로 채워줄 수 있다. 
    -> 헷갈리면 안 되는게, 예를 들어 "{:010d}".format(123)이 존재한다고 가정하면, 123은 3칸을 차지하므로, 앞에 7칸만이 0으로 채워진다.

--> 지금까지 d를 붙인 것은 정수이기 때문에 d를 붙인 것이며, f를 붙임으로써 부동소수점 표현도 가능하게 된다. 

In [29]:
output_a = "{:+d}".format(-33)
output_b = "{:+d}".format(33)
print(output_a)
print(output_b)

output_c = "{:5d}".format(-33)
output_d = "{:10d}".format(33)
print(output_c)
print(output_d)

output_e = "{:+05d}".format(-33)
output_f = "{:010d}".format(33)
print(output_e)
print(output_f)

-33
+33
  -33
        33
-0033
0000000033
